### PR TITLE
Add CID to mmpose

### PR DESCRIPTION
**Pull request by kennethwdk**
## Motivation

Integrate CID (CVPR2022) into mmpose.

## Modification

Modify dataset to add 'bbox' and 'area' into db_rec
Modify pipeline to support transformation of 'bbox' and add CID target generator 
Add CID detector and head and configs. The code is adapted from [here](https://github.com/kennethwdk/CID)

The log and weights re-trained by mmpose are provided at [here](https://1drv.ms/u/s!AhpKYLhXKpH7g4dWSAsyog1FFlRcqA?e=b1iOr2)

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by jin-s13** [comment]
Thanks a lot for your contributions! We will review the codes ASAP.

**Comment by jin-s13** [comment]
Could you please follow the instructions [here](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to setup the pre-commit? @kennethwdk

**Comment by jin-s13** [comment]
BTW, would you mind adding some unit tests for CID?
You may refer to https://github.com/open-mmlab/mmpose/blob/master/tests/test_models/test_bottom_up_forward.py as an example.

File: mmpose/models/heads/cid_head.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        instance_locations = instance_locations
+        relative_coords = instance_locations.reshape(-1, 1, 2) - locations.reshape(1, -1, 2)
+        relative_coords = relative_coords.permute(0, 2, 1).float()
+        relative_coords = (relative_coords / 32).to(dtype=global_features.dtype)
+        relative_coords = relative_coords.reshape(n_inst, 2, H, W)
+        input_feats = torch.cat((input_feats, relative_coords), dim=1)
+        mask = self.conv(input_feats).sigmoid()
+        return global_features * mask
+
+class FocalLoss(nn.Module):
```
Comment (jin-s13) [inline mmpose/models/heads/cid_head.py:273]: How about moving these Loss to `mmpose/models/losses`?
For example, https://github.com/open-mmlab/mmpose/blob/master/mmpose/models/losses/mse_loss.py

**Comment by kennethwdk** [comment]
Hello, I have created new commits to add the forward test, moving the losses and use pre-commit to check code styles.

File: mmpose/models/heads/cid_head.py (hunk 1 above)
Comment (kennethwdk) [reply mmpose/models/heads/cid_head.py]: I move the FocalLoss (renamed as FocalHeatmapLoss) to mmpose/models/losses module, because this loss can also be used by other projects. I did not move the contrastive loss because I think it may be rarely used by other projects and moving it will cause redundancy.

**Comment by Ben-Louis** [comment]
Hi, we retrain the CID model and get better performance.

| Arch                                          | Input Size |  AP   | AP<sup>50</sup> | AP<sup>75</sup> |  AR   | AR<sup>50</sup> | 
| :-------------------------------------------- | :--------: | :---: | :-------------: | :-------------: | :---: | :-------------: |
| CID |  512x512   | 0.702 |      0.887      |      0.768      | 0.755 |      0.926      |

The corresponding config file has been updated

**Comment by OpenMMLab-Assistant005** [comment]
Hi @kennethwdk ！First of all, we want to express our gratitude for your significant PR in the MMPose project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
